Data Converter updates (#1932)
temporalio/documentation: docs-src/java/how-to-use-a-custom-payload-codec-in-java.md

---
id: how-to-use-a-custom-payload-codec-in-java
title: How to use a custom Payload Codec in Java
sidebar_label: Custom Payload Codec
description: Create a custom implementation of `PayloadCodec` and use it in `CodecDataConverter` to set a custom Data Converter.
tags:
  - Java
  - developer-guide
---

**Create a custom Payload Codec**

> Create a custom implementation of [`PayloadCodec`](https://www.javadoc.io/static/io.temporal/temporal-sdk/1.18.1/io/temporal/payload/codec/PayloadCodec.html) and use it in [`CodecDataConverter`](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/common/converter/CodecDataConverter.html)) to set a custom Data Converter.

Reviewer: We need to unify the style of the links to reference docs.
I'm not sure what is our style.
If it's "latest", all links should be not to tags, but to the "latest" state.
If we reference some specific tag (which has it's own merit that such links will never die), it should be tag-based urls for all languages on all pages.

Author: yes; I'm checking all links and correcting them to point to latest.

The Payload Codec does byte-to-byte conversion and must be set with a Data Converter.

Define custom encryption/compression logic in your `encode` method and decryption/decompression logic in your `decode` method.

The following example from the [Java encryption sample](https://github.com/temporalio/samples-java/blob/main/src/main/java/io/temporal/samples/encryptedpayloads/CryptCodec.java) shows how to implement encryption and decryption logic on your payloads in your `encode` and `decode` methods.

```java
class YourCustomPayloadCodec implements PayloadCodec {

>   //...

Reviewer: ```suggestion

```

  // See the linked sample for details on the methods called here.
  @NotNull
  @Override
  public List<Payload> encode(@NotNull List<Payload> payloads) {
    return payloads.stream().map(this::encodePayload).collect(Collectors.toList());
  }

  @NotNull
  @Override
  public List<Payload> decode(@NotNull List<Payload> payloads) {
    return payloads.stream().map(this::decodePayload).collect(Collectors.toList());
  }

  private Payload encodePayload(Payload payload) {

>     String keyId = getKeyId();

Reviewer: If you carry usages of these methods but don't want to carry an implementation, you should as least save the method definition with an empty body and add a comment that "this should do this and that and a specific implementation is not important for this sample."

Author: Yep; adding comments inline for these

>     String keyId = getKeyId(); // Get a fixed key Id from the getKeyId method. See the sample for details.

Reviewer: I already wrote a comment here, that getKeyId() and getKey() should be saved in this snippet on the level of signatures and a short doc. An implementation can be omitted as not important for the purpose of the sample and there should be a note about it. Otherwise, it's just an incorrect partial code.

Author: yeah. The idea here was to just show specific implementation example and direct the reader to the sample. But it makes sense to add them in, so updated!

    SecretKey key = getKey(keyId); // Get an AES key defined in the getKey method. See sample for details.

    byte[] encryptedData;
    try {
      encryptedData = encrypt(payload.toByteArray(), key);
    } catch (Throwable e) {
      throw new DataConverterException(e);
    }
    // Apply metadata to the encoded payload that you can verify in your decode method before decoding.
    // See the sample for details on the metadata values set.
    return Payload.newBuilder()

>         .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, METADATA_ENCODING)

Reviewer: This piece of code is copy-pasted, the costants are still used, but why the constants themselves are not transferred together?

Author: Added the constants at the top, like in the sample. It looks cleaner.

Author: and added a comment inline for context.

        .putMetadata(METADATA_ENCRYPTION_CIPHER_KEY, METADATA_ENCRYPTION_CIPHER)
        .putMetadata(METADATA_ENCRYPTION_KEY_ID_KEY, ByteString.copyFromUtf8(keyId))
        .setData(ByteString.copyFrom(encryptedData))
        .build();
  }

  private Payload decodePayload(Payload payload) {
    // Verify the incoming encoded payload metadata before applying decryption.
    if (METADATA_ENCODING.equals(
        payload.getMetadataOrDefault(EncodingKeys.METADATA_ENCODING_KEY, null))) {
      String keyId;
      try {
        keyId = payload.getMetadataOrThrow(METADATA_ENCRYPTION_KEY_ID_KEY).toString(UTF_8);
      } catch (Exception e) {
        throw new PayloadCodecException(e);
      }
      SecretKey key = getKey(keyId); // Gets the AES key defined in the getKey method which was used for the encryption. See sample for details.

      byte[] plainData;
      Payload decryptedPayload;

      try {
        plainData = decrypt(payload.getData().toByteArray(), key);
        decryptedPayload = Payload.parseFrom(plainData);
        return decryptedPayload;
      } catch (Throwable e) {
        throw new PayloadCodecException(e);
      }
    } else {
      return payload;
    }
  }

>   //...

Reviewer: ```suggestion
```

---
id: how-to-use-a-custom-payload-codec-in-java
title: How to use a custom Payload Codec in Java
sidebar_label: Custom Payload Codec
description: Create a custom implementation of the `PayloadCodec` and use it in a `CodecDataConverter` to set a custom Data Converter.
tags:
  - Java
  - developer-guide
---

**Create a custom Payload Codec**

Create a custom implementation of the [`PayloadCodec`](https://www.javadoc.io/static/io.temporal/temporal-sdk/1.18.1/io/temporal/payload/codec/PayloadCodec.html) and use it in a `CodecDataConverter` to set a custom Data Converter.

The Paylod Codec does byte to byte conversion, and must be set with a Data Converter to do the conversion to bytes and values.

Define custom encryption/compression logic in your `encode` method, and decryption/decompression logic in your `decode` method.

```java
public class YourCustomPayloadCodec implements PayloadCodec {

>      @NotNull

Reviewer: allignment

    @Override
    public List<Payload> encode(@NotNull List<Payload> payloads) {
        //your encryption/compression logic
    }

    @NotNull
    @Override
    public List<Payload> decode(@NotNull List<Payload> payloads) {
        //your decryption/decompression logic
    }

> ```

Reviewer: unbalanced braces

> You can also create a remote HTTP server (called Codec Server) to run the encryption and decryption through the custom `PayloadCodec`, and expose endpoints that you can use with WebUI and tctl to see decrypted data.

Reviewer: We call it Remote Data Encoder usually

description: Create a custom implementation of `PayloadCodec` and use it in `CodecDataConverter` to set a custom Data Converter.
tags:
  - Java
  - developer-guide
---

**Create a custom Payload Codec**

Create a custom implementation of [`PayloadCodec`](https://www.javadoc.io/static/io.temporal/temporal-sdk/1.18.1/io/temporal/payload/codec/PayloadCodec.html) and use it in [`CodecDataConverter`](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/common/converter/CodecDataConverter.html)) to set a custom Data Converter.

The Payload Codec does byte-to-byte conversion and must be set with a Data Converter to do the conversion to bytes and values.

Define custom encryption/compression logic in your `encode` method, and decryption/decompression logic in your `decode` method.

```java
public class YourCustomPayloadCodec implements PayloadCodec {
    @NotNull
    @Override
    public List<Payload> encode(@NotNull List<Payload> payloads) {
        //your encryption/compression logic
    }

    @NotNull
    @Override
    public List<Payload> decode(@NotNull List<Payload> payloads) {
        //your decryption/decompression logic
    }
    //...
}

> ```

Author: added the logic in, because this is an example. How we do this will potentially change soon (https://github.com/temporalio/documentation-samples-go)

> You can also create a remote HTTP server (called Codec Server) to run the encryption and decryption through the custom `PayloadCodec`, and expose endpoints that you can use with WebUI and tctl to see decrypted data.

Reviewer 2: ```suggestion
You can also create a remote HTTP server (called a codec server) to run the encryption and decryption through the custom `PayloadCodec`, and expose endpoints that you can use with Web UI and tctl to see decrypted data.
```
But see Dmitry's comment and apply any change both here and near the end of `docs-src/go/how-to-use-a-custom-payload-codec-in-go.md`.

**Set Data Converter to use custom Payload Codec**

Use `CodecDataConverter` with an instance of a Data Converter and the custom `PayloadCodec` in the `WorkflowClient` options that you use in your Worker process and to start your Workflow Executions.

> For example, to set a custom `PayloadCodec` implementation with the `DefaultDataConverter`, use:

Reviewer 2: ```suggestion
For example, to set a custom `PayloadCodec` implementation with the `DefaultDataConverter`, use the following code:
```

```java
WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();

>   // client that can be used to start and signal workflows

Reviewer 2: ```suggestion
  // Client that can be used to start and signal Workflows
```
